### PR TITLE
HDDS-6551. Introduce StatefulService in scm

### PR DESCRIPTION
**Pull request by siddhantsangwan**
## What changes were proposed in this pull request?

A `StatefulService` is an `SCMService` with an API to persist/read configuration to RocksDB and replicate it through RATIS. The configuration would be provided as a Protobuf message. This change includes an interface `StatefulServiceStateManager` and its implementation `StatefulServiceStateManagerImpl`.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6551

## How was this patch tested?

Added a unit test.

### CODE REVIEW

**Comment by siddhantsangwan** [comment]
@nandakumar131 @lokeshj1703 can you please review?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/StatefulServiceStateManagerImpl.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      this.transactionBuffer = dbTransactionBuffer;
+      return this;
+    }
+
+    public Builder setRatisServer(final SCMRatisServer ratisServer) {
+      scmRatisServer = ratisServer;
+      return this;
+    }
+
+    public StatefulServiceStateManager build() {
```
Comment (nandakumar131) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/StatefulServiceStateManagerImpl.java:112]: Add
```
Preconditions.checkNotNull(statefulServiceConfig);
Preconditions.checkNotNull(transactionBuffer);
Preconditions.checkNotNull(scmRatisServer);
```

**Review by nandakumar131 — COMMENTED** [review]
Overall the change looks good.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/StatefulServiceStateManagerImpl.java (hunk 1 above)
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/StatefulServiceStateManagerImpl.java:107]: Can we exclude the not null check for `scmRatisServer`? `SCMHAInvocationHandler` takes care of the null case by calling the `invokeLocal` method.

**Review by nandakumar131 — APPROVED** [review]

**Pull request merged**